File: src/CloudGlyph/Assets/Docs/scripts/validate-katex.py
"""
validate-katex.py

Deterministic validation of KaTeX math delimiters in Markdown wiki content.
Use this in the skill's Review phase instead of relying on the agent's eye.

The scanner ignores content inside fenced code blocks, because `$` there is usually
not math (e.g. C# string interpolation `$"..."` in code) and produced false positives.

Checks:
  ERROR  odd count of `$$` block-math delimiters across a file (unbalanced $$ ... $$)
  ERROR  odd count of remaining `$` after removing $$ blocks (unbalanced inline $...$)
  ERROR  single-line display math `$$expr$$` (open+close on the same line) — the CloudGlyph
         viewer's block parser only recognizes multi-line `$$\n…\n$$`, so these never render
  WARN   a CJK character used directly in math outside a `\text{...}` group — KaTeX lacks CJK
         glyphs and only warns; wrap prose in `\text{...}` or keep it outside the formula
  WARN   `$"` seen outside code fences (a C#-style interpolated string leaked into prose)
  WARN   `$ ` or ` $` (dollar adjacent to a space) — probably currency, not math

Usage:
    python validate-katex.py                    # auto-detect content root
    python validate-katex.py --lang en
    python validate-katex.py <dir-or-file>

Exit code: 0 = clean, 1 = at least one ERROR.
"""

import argparse
import re
import sys
from pathlib import Path

FENCE_RE = re.compile(r"^```\w*\s*$")
DOLLAR_DOLLAR_RE = re.compile(r"\$\$")
DOLLAR_RE = re.compile(r"\$(?!\$)")  # single $ not followed by another $
CJK_RE = re.compile(r"[一-鿿]")
SAME_LINE_DISPLAY_RE = re.compile(r"\$\$.+?\$\$")
TEXT_CMD = "\\text{"


def _cjk_outside_text(math: str) -> str:
    """Characters in a math snippet that are CJK and NOT inside a \\text{...} group."""
    found = []
    i, n = 0, len(math)
    while i < n:
        if math.startswith(TEXT_CMD, i):
            depth, j = 1, i + len(TEXT_CMD)
            while j < n and depth > 0:
                if math[j] == "{":
                    depth += 1
                elif math[j] == "}":
                    depth -= 1
                j += 1
            i = j
            continue
        if math[i] == "\\" and i + 1 < n and math[i + 1].isalpha():
            i += 2
            continue
        if CJK_RE.match(math[i]):
            found.append(math[i])
        i += 1
    return "".join(found)


def _math_spans(line: str):
    """Yield the content of every $...$ / $$...$$ span on one prose line."""
    i, n = 0, len(line)
    while i < n:
        if line[i] != "$":
            i += 1
            continue
        j = i + 1
        dbl = j < n and line[j] == "$"
        if dbl:
            j += 1
        end = line.find("$$", j) if dbl else line.find("$", j)
        if end == -1:
            break
        yield line[j:end]
        i = end + (2 if dbl else 1)


def split_fences(lines):
    """Return (code_fenced, prose) tuples: (in_code, line)."""
    out = []
    in_code = False
    for line in lines:
        if FENCE_RE.match(line.strip()):
            in_code = not in_code
            out.append((True, line))
            continue
        out.append((in_code, line))
    return out


def validate_file(lines):
    issues = []
    fenced = split_fences(lines)
    prose = "\n".join(line for in_code, line in fenced if not in_code)
    # re-fence aware prose line indices for reporting
    prose_line_numbers = [i + 1 for i, (in_code, _) in enumerate(fenced) if not in_code]

    # block math: $$ ... $$
    dd = len(DOLLAR_DOLLAR_RE.findall(prose))
    if dd % 2 != 0:
        issues.append(("ERROR", f"odd count of '$$' block delimiters ({dd}) — unbalanced $$ ... $$"))

    # remove complete $$...$$ spans, then count leftover single $
    stripped = DOLLAR_DOLLAR_RE.sub("", prose)
    single = len(DOLLAR_RE.findall(stripped))
    if single % 2 != 0:
        issues.append(("ERROR", f"odd count of inline '$' delimiters ({single}) — unbalanced $...$"))

    # heuristic warnings on prose (conservative: a `$` followed by a digit looks like
    # currency, not math; `$` before/after whitespace is normal inline-math and NOT flagged)
    for m in re.finditer(r'\$"', prose):
        issues.append(("WARN", "C#-style $\" interpolated string outside a code fence"))
    for ln in prose_line_numbers:
        line = fenced[ln - 1][1] if 0 <= ln - 1 < len(fenced) else ""
        # currency: `$` + digits + NOT a math continuation (math like `$2S$` has a letter
        # right after the digits and must not be flagged)
        if re.search(r"\$[\d]+(?=[\s.,;:!?]|$)", line):
            issues.append(("WARN", f"line {ln}: '$' followed by digits — likely currency, verify it is math"))

    # Renderer-compatibility checks (the CloudGlyph viewer = AvalonMarkdown block parser):
    for i, (in_code, line) in enumerate(fenced):
        if in_code:
            continue
        ln = i + 1
        if SAME_LINE_DISPLAY_RE.search(line):
            issues.append(("ERROR",
                f"line {ln}: single-line display math '$$…$$' — open and close on the same line; "
                "the viewer only renders display math as a standalone multi-line '$$' block"))
        for math in _math_spans(line):
            bare = _cjk_outside_text(math)
            if bare:
                issues.append(("WARN",
                    f"line {ln}: CJK characters {bare!r} used in math outside \\text{{}} — "
                    "wrap them in \\text{...} or move the prose out of the formula"))

    return issues


def main():
    for stream in (sys.stdout, sys.stderr):
        try:
            stream.reconfigure(encoding="utf-8")
        except (AttributeError, ValueError):
            pass

    parser = argparse.ArgumentParser(description="Validate KaTeX delimiters in wiki Markdown")
    parser.add_argument("path", nargs="?", help="file or directory to scan (default: auto-detect content root)")
    parser.add_argument("--lang", help="only scan this language subdir (e.g. en, zh)")
    args = parser.parse_args()

    if args.path:
        target = Path(args.path)
    else:
        content = Path(__file__).resolve().parent.parent.parent / "content"
        target = content / args.lang if args.lang else content

    files = [target] if target.is_file() else sorted(target.rglob("*.md"))
    total_errors = total_warns = 0

    for f in files:
        try:
            lines = f.read_text(encoding="utf-8").split("\n")
        except OSError as e:
            print(f"[katex] ERROR reading {f}: {e}")
            total_errors += 1
            continue
        for level, msg in validate_file(lines):
            if level == "ERROR":
                total_errors += 1
                print(f"{f}: ERROR {msg}")
            else:
                total_warns += 1
                print(f"{f}: WARN  {msg}")

    print(f"[katex] scanned {len(files)} file(s); {total_errors} error(s), {total_warns} warning(s)")
    sys.exit(1 if total_errors else 0)


if __name__ == "__main__":
    main()

File: src/CloudGlyph/Assets/Docs/scripts/validate-katex.js
#!/usr/bin/env node
/**
 * validate-katex.js — REAL-engine validation of KaTeX math.
 *
 * Every `$...$` / `$$...$$` expression in prose is rendered with the actual
 * `katex` package (throwOnError). Invalid math throws with the exact error —
 * ground truth, not a heuristic. Content inside code fences is skipped
 * (`$` there is usually C# interpolation, not math).
 *
 * Usage:
 *   node validate-katex.js                    # auto-detect content root
 *   node validate-katex.js <dir-or-file>      # explicit path
 *
 * Requires: `npm install` in this directory (see package.json).
 * Exit code: 0 = clean, 1 = at least one render error.
 */
const fs = require("fs");
const path = require("path");
const katex = require("katex");

const FENCE = /^```\w*\s*$/;
const BLOCK = /\$\$(.*?)\$\$/gs; // $$ ... $$ (may span lines)
const INLINE = /\$(.+?)\$/g; // $ ... $ (single line)

function isMathSpan(open, close, s, start) {
  // crude guard: a `$` that is part of `$$` is handled by BLOCK pass first
  return true;
}

function walk(dir, out = []) {
  for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) walk(p, out);
    else if (e.name.endsWith(".md")) out.push(p);
  }
  return out;
}

function renderAll(text) {
  const out = [];
  // display math first ($$ ... $$)
  let rest = text;
  for (const m of text.matchAll(BLOCK)) {
    out.push({ display: true, math: m[1] });
  }
  rest = text.replace(BLOCK, "  ");
  // inline math ($ ... $)
  for (const m of rest.matchAll(INLINE)) {
    out.push({ display: false, math: m[1] });
  }
  return out;
}

const CJK = /[一-鿿]/;
const TEXT_CMD = "\\text{";
const SAME_LINE_DISPLAY = /\$\$.+?\$\$/;

function cjkOutsideText(math) {
  // CJK characters in a math snippet that are NOT inside a \text{...} group.
  const found = [];
  let i = 0;
  while (i < math.length) {
    if (math.startsWith(TEXT_CMD, i)) {
      let depth = 1, j = i + TEXT_CMD.length;
      while (j < math.length && depth > 0) {
        if (math[j] === "{") depth++;
        else if (math[j] === "}") depth--;
        j++;
      }
      i = j;
      continue;
    }
    if (math[i] === "\\" && i + 1 < math.length && /[A-Za-z]/.test(math[i + 1])) {
      i += 2;
      continue;
    }
    if (CJK.test(math[i])) found.push(math[i]);
    i++;
  }
  return found.join("");
}

function mathSpans(line) {
  const spans = [];
  let i = 0;
  while (i < line.length) {
    if (line[i] !== "$") { i++; continue; }
    let j = i + 1;
    const dbl = j < line.length && line[j] === "$";
    if (dbl) j++;
    const end = dbl ? line.indexOf("$$", j) : line.indexOf("$", j);
    if (end === -1) break;
    spans.push(line.slice(j, end));
    i = end + (dbl ? 2 : 1);
  }
  return spans;
}

function main() {
  const arg = process.argv[2];
  const target = arg ? path.resolve(arg) : path.resolve(__dirname, "../content");
  const files = fs.statSync(target).isFile() ? [target] : walk(target);
  let errors = 0, warnings = 0, rendered = 0;

  for (const f of files) {
    const lines = fs.readFileSync(f, "utf8").split("\n");
    let inCode = false;
    const proseLines = []; // {n, line}
    for (let i = 0; i < lines.length; i++) {
      if (FENCE.test(lines[i].trim())) {
        inCode = !inCode;
        continue;
      }
      if (!inCode) proseLines.push({ n: i + 1, line: lines[i] });
    }
    const proseText = proseLines.map((p) => p.line).join("\n");
    for (const { display, math } of renderAll(proseText)) {
      rendered++;
      try {
        katex.renderToString(math, { throwOnError: true, displayMode: display });
      } catch (e) {
        errors++;
        console.log(`${f}: ERROR katex (${display ? "$$" : "$"}${math}${display ? "$$" : "$"}): ${e.message}`);
      }
    }
    // Renderer-compatibility checks (CloudGlyph viewer = AvalonMarkdown block parser):
    for (const { n, line } of proseLines) {
      if (SAME_LINE_DISPLAY.test(line)) {
        errors++;
        console.log(`${f}:${n}: ERROR single-line display math '$$…$$' (open+close on one line) — ` +
          "the viewer only renders display math as a standalone multi-line '$$' block");
      }
      for (const math of mathSpans(line)) {
        const bare = cjkOutsideText(math);
        if (bare) {
          warnings++;
          console.log(`${f}:${n}: WARN CJK characters ${JSON.stringify(bare)} in math outside \\text{} — ` +
            "wrap in \\text{...} or move the prose out of the formula");
        }
      }
    }
  }
  console.log(`[katex-real] rendered ${rendered} expression(s); ${errors} error(s), ${warnings} warning(s)`);
  process.exit(errors ? 1 : 0);
}

main();
